File: cpp/n160_intersectionNode.cpp
// https://leetcode-cn.com/problems/intersection-of-two-linked-lists/
// Definition for singly-linked list.
#include <iostream>
#include <vector>
#include <map>
using namespace std;

struct ListNode {
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(NULL) {}
};

class Solution {
public:
    ListNode *getIntersectionNode(ListNode *headA, ListNode *headB) {
        ListNode* p_a = headA;
        ListNode* p_b = headB;
        while (p_a != p_b) {
            if (p_a == NULL) p_a = headB;
            else p_a = p_a -> next;
            if (p_b == NULL) p_b = headA;
            else p_b = p_b -> next;
        }
        return p_a;
    }
    
    ListNode *getIntersectionNode_a(ListNode *headA, ListNode *headB) {
        ListNode* p_a = headA;
        ListNode* p_b = headB;
        map <ListNode*, bool> node_map;
        while(p_a != NULL){
            node_map.insert(pair<ListNode*, bool>(p_a, true));
            p_a = p_a -> next;
        }
        while(p_b != NULL){
            auto iter = node_map.find(p_b);
            if (iter != node_map.end()) return p_b;
            p_b = p_b -> next;
        }
        return NULL;
    }
};

int main(){
    return 0;
}

File: python/n160_IntersectionNode.py
#https://leetcode-cn.com/problems/intersection-of-two-linked-lists/
# Definition for singly-linked list.
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution:
    def getIntersectionNode(self, headA: ListNode, headB: ListNode) -> ListNode:
        if headA is None or headB is None:
            return None
        p_a = headA
        p_b = headB
        while p_a != p_b:
            if p_a is not None:
                p_a = p_a.next
            else: p_a = headB
            if p_b is not None:
                p_b = p_b.next
            else: p_b = headA
        return p_a
    
    def getIntersectionNode_a(self, headA: ListNode, headB: ListNode) -> ListNode:
        node_map = {}
        if headA is None or headB is None:
            return None
        p_a = headA
        p_b = headB
        while p_a is not None:
            node_map[p_a] = True
            p_a = p_a.next
        while p_b is not None:
            if p_b in node_map:
                return p_b
            p_b = p_b.next

if __name__ == "__main__":
    pass